gitpod-io/gitpod
# use sync.Cond to notify about gitpod config changes

## Description


There was some kind of deadlock while one listener was trying to read config changes. I rewrote gitpod config service to use conditional variable to announce changes in gitpod config to avoid synchronisation over map of listeners. I changed clients as well that they don't consider nil config as closed channel and stop reading.

## Related Issue(s)

fix #5929

## How to test


- Start a workspace.
- Change .gitpod.yml to add port ranges ignoring some ports like in #5929
- Try to start ports inside and outside of such port ranges and verify that ignore is respected.
- Change onOpen action for this range, try to restart service on the port and see that changes are applied.

## Release Notes

```release-note
NONE
```
